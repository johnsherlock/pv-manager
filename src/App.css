.App {
  text-align: center;
}

.App-logo {
  height: 40vmin;
  pointer-events: none;
}

@media (prefers-reduced-motion: no-preference) {
  .App-logo {
    animation: App-logo-spin infinite 20s linear;
  }
}

.App-header {
  background-color: #282c34;
  min-height: 100vh;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  font-size: calc(10px + 2vmin);
  color: white;
}

.App-link {
  color: #61dafb;
}

@keyframes App-logo-spin {
  from {
    transform: rotate(0deg);
  }
  to {
    transform: rotate(360deg);
  }
}

/* table {
  width: 100%;
  border-collapse: collapse;
  overflow-y: scroll;
  max-height:600px;
}
thead,
tfoot {
    display: table;
    width: 100%;
}
tbody {
    display: block;
    overflow-y: scroll;
    max-height: 600px;
} */

.grid-container {
  display: grid;
  grid-template-rows: auto 1fr;
  height: 100%;
}

.table-container {
  grid-row: 1;
  min-height: 250px;
  max-height: 330px;
  overflow-y: scroll;
}
.display {
  display: grid;
}
.hide {
  display: none;
}

.table-header {
  flex-direction: row;
  background-color: #eee;
  font-weight: 600;
}

.table-footer {
  flex-direction: row;
  background-color: #ddd;
  font-weight: 600;
}

.table-row {
  display: flex;
  flex-direction: row;
  
}

.table-primary {
  background-color: #E1F4FF;
}

.table-cell {
  flex: 1;
  padding: 0px;
}

.chart {
  display: grid;
  grid-row: 2;
  height: 480px;
}

.navigationButton, .date {
  display: inline-block;
  float: left;
  padding-right: 20px;
  cursor: pointer;
}

.navigation::after {
  content: "";
  clear: both;
  display: table;
}