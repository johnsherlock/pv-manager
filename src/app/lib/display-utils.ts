export const isTouchScreen = () => 'ontouchstart' in window || navigator.maxTouchPoints > 0;