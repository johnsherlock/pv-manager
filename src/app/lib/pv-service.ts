import axios from 'axios';
import { EddiData } from '../../shared/eddi-data';
import { mapEddiDataToMinutePVData } from '../../shared/energy-utils';
import { MinutePVData, Totals } from '../../shared/pv-data';

export interface DailyEnergyUsageProps {
  data: MinutePVData[];
}

export const getPVDataForDate = async (formattedTargetDate: string): Promise<MinutePVData[]> => {
  console.log(`Retrieving per minute data for ${formattedTargetDate}`);
  const url = `https://jmcjm1731b.execute-api.eu-west-1.amazonaws.com/prod/minute-data?date=${formattedTargetDate}`;
  try {
    const response = await axios.get(url);
    const data = await Promise.resolve(response.data as EddiData[]);
    return data.map(mapEddiDataToMinutePVData);
  } catch (error) {
    console.log(`Error retrieving data for ${formattedTargetDate}`, error);
    document.body.style.cursor = 'auto';
    return [];
  };
};

export const getPVTotalsForRange = async (startDate: string, endDate: string): Promise<Totals[]> => {
  console.log(`Retrieving PV summary data for range ${startDate} - ${endDate}`);
  const url = `https://jmcjm1731b.execute-api.eu-west-1.amazonaws.com/prod/aggregate-historical-data?serialNumber=21494842&startDate=${startDate}&endDate=${endDate}`;
  try {
    const response = await axios.get(url);
    return await Promise.resolve(response.data as Totals[]);
  } catch (error) {
    console.log('Error retrieving totals for range', error);
    document.body.style.cursor = 'auto';
    return [];
  };
};