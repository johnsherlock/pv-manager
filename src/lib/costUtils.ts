import * as numUtils from './numUtils';

// export discounted rates
// export const dayRate = 0.4330;
// export const peakRate = 0.5289;
// export const nightRate = 0.3182;

// export 100% standard rates from bill
export const dayRate = 0.4673;
export const peakRate = 0.5709;
export const nightRate = 0.3434;

// export calculated 100% standard rates
// export const dayRate = 0.5094;
// export const peakRate = 0.6222;
// export const nightRate = 0.3742;

export const exportRate = 0.1850;

export const discount = 1 - 0.15;
export const standingCharge = 0.7066;
export const hourlyStandingCharge = standingCharge / 24;
export const vatRate = 1.09;

export const formatToEuro = (amount: number) => (amount ? `â‚¬${amount.toFixed(2)}` : '');

export const calculateHourlyNetCostAtStandardRates = (hour = 0, dow: string, joules: number) => {
  if (joules) {
    let multiplier = dayRate;

    if ((hour >= 0 && hour <= 8) || hour === 23) {
      multiplier = nightRate;
    } else if (hour >= 17 && hour < 19) {
      multiplier = peakRate;
    }

    if (dow === 'Sat' && hour >= 9 && hour < 17) {
      multiplier = 0;
    }

    const kWh = numUtils.convertJoulesToKwh(joules);
    const cost = kWh * multiplier;

    return numUtils.formatDecimal(cost);
  }
  return 0;
};

export const calculateDiscountedHourlyGrossCost = (hour: number = 0, dow: string, joules: number) => {
  const netCostAtStandardRates = calculateHourlyNetCostAtStandardRates(hour, dow, joules);
  const grossCost = (netCostAtStandardRates * discount) * vatRate;
  return numUtils.formatDecimal(grossCost);
};

export const calculateHourlyGrossCostIncStdChgAndDiscount = (hour: number = 0, dow: string, joules: number) => {
  const netCostAtStandardRates = calculateHourlyNetCostAtStandardRates(hour, dow, joules);
  const grossCost = ((netCostAtStandardRates * discount) + hourlyStandingCharge) * vatRate;
  return numUtils.formatDecimal(grossCost);
};

export const calculateSaturdaySaving = (hour = 0, dow: string, joules: number) => {
  if (joules && dow === 'Sat' && hour >= 9 && hour <= 17) {
    const kWh = numUtils.convertJoulesToKwh(joules);
    const netSavingAtStandardRates = kWh * dayRate;
    const grossSavingAtDiscountedRates = (netSavingAtStandardRates * discount) * vatRate;
    return numUtils.formatDecimal(grossSavingAtDiscountedRates);
  }
  return 0;
};

export const calculateDiscountedGrossCostExcludingStdChg = (netCostAtStandardRate: number) => {
  const grossCost = (netCostAtStandardRate * discount) * vatRate;
  return numUtils.formatDecimal(grossCost);
};

export const calculateGrossCostIncStandingCharges = (netCost: number) => {
  const grossCost = ((netCost * discount) + hourlyStandingCharge) * vatRate;
  return numUtils.formatDecimal(grossCost);
};

export const calculateExportValue = (joules: number) => {
  if (joules) {
    const kWh = numUtils.convertJoulesToKwh(joules);
    return numUtils.formatDecimal(kWh * exportRate);
  }
  return 0;
};
