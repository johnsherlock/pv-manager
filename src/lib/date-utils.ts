import moment from 'moment-timezone';
import { BasePVData } from '../model/pv-data';

export const formatDate = (date: moment.Moment): string => moment(date).startOf('day').format('YYYY-MM-DD');

export const dawnOfTime = moment(new Date('2023-01-20')).startOf('day');

export const toMoment = (pvData: BasePVData, locale?: string): moment.Moment => {
  return moment({
    year: pvData.year,
    month: pvData.month - 1,
    date: pvData.dayOfMonth,
    hour: pvData.hour,
    minute: pvData.minute,
  });
};

export const getFormattedTime = (pvData: BasePVData, locale?: string): string => {
  const time = toMoment(pvData, locale);
  return time.format('HH:mm');
};