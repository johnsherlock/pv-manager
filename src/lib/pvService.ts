import axios from 'axios';

export interface DailyEnergyUsageProps {
  data: any;
}

export interface HourlyUsageData {
  'yr': number;
  'mon': number;
  'dom': number;
  'dow': 'Sun' | 'Mon' | 'Tues' | 'Wed' | 'Thurs' | 'Fri' | 'Sat';
  'hr': number;
  'min': number;
  'imp'?: number;
  'gep'?: number;
  'pect1'?: number;
  'pect2'?: number;
  'hsk?': number;
  'v1'?: number;
  'frq'?: number;
}

export const getHourlyUsageDataForDate = async (formattedTargetDate: string): Promise<HourlyUsageData[]> => {
  console.log(`Retrieving data for ${formattedTargetDate}`);
  const url = `http://localhost:3001/hour-data?date=${formattedTargetDate}`;
  try {
    const response = await axios.get(url);
    return await Promise.resolve(response.data.U21494842 as HourlyUsageData[]);
  } catch (error) {
    console.log('Error retrieving remote data', error);
    document.body.style.cursor = 'auto';
    return [];
  };
};