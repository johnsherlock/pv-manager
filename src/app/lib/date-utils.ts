import moment from 'moment';
import { CalendarScale, DateRange, FormattedDateRange } from './state-utils';
import { BasePVData } from '../../shared/pv-data';

export const formatDate = (date: moment.Moment): string => moment(date).startOf('day').format('YYYY-MM-DD');

export const dawnOfTime = moment(new Date('2023-01-20')).startOf('day');

export const toMoment = (pvData: BasePVData, locale?: string): moment.Moment => {
  return moment({
    year: pvData.year,
    month: pvData.month - 1,
    date: pvData.dayOfMonth,
    hour: pvData.hour,
    minute: pvData.minute,
  });
};

export const getFormattedTime = (pvData: BasePVData, locale?: string): string => {
  const time = toMoment(pvData, locale);
  return time.format('HH:mm');
};

export const getDateRange = (scale: CalendarScale, date: Date): DateRange => {

  switch (scale) {
    case 'week':
    case 'month':
    case 'year': {
      const m = moment(date);
      const start = moment(m.startOf(scale).toDate());
      const end = moment(m.endOf(scale).toDate());
      return { startDate: start, endDate: end };
    }
    default:
      return { startDate: moment(date), endDate: (moment(date)) };
  }
};

export const formatDateRange = (dateRange: DateRange): FormattedDateRange => {
  return {
    startDate: formatDate(dateRange.startDate),
    endDate: formatDate(dateRange.endDate),
  };
};