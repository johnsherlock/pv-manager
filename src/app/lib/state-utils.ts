import moment from 'moment';
import * as dateUtils from './date-utils';
import { MinutePVData, Totals } from '../../shared/pv-data';
import { Scale, View } from '../energy-usage-line-graph';

export type CalendarScale = 'day' | 'week' | 'month' | 'year' | 'custom';

export interface AppState {
  intervalId?: number | null;
  today: moment.Moment;
  selectedDate: moment.Moment;
  startDate?: moment.Moment;
  endDate?: moment.Moment;
  formattedSelectedDate: string;
  pvDataCache: Map<string, MinutePVData[]>;
  totals: Map<string, Totals>;
  energyUsageLineGraphScale: Scale;
  energyUsageLineGraphView: View;
  calendarScale: CalendarScale;
}

export const initialState = (): AppState => {
  const today: moment.Moment = moment().startOf('day');
  return {
    today: today,
    selectedDate: today,
    formattedSelectedDate: dateUtils.formatDate(today),
    pvDataCache: new Map(),
    totals: new Map(),
    energyUsageLineGraphScale: 'hour',
    energyUsageLineGraphView: 'line',
    calendarScale: 'day',
  };
};