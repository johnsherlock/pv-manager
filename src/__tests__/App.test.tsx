import { render, act, waitFor } from '@testing-library/react';
import moment from 'moment';
import React from 'react';
import App from '../App';
import * as pvService from '../lib/pvService';

jest.mock('../lib/pvService', () => ({
  getHourlyUsageDataForDate: jest.fn().mockResolvedValue([
    { yr: 2023, mon: 1, dom: 31, imp: 1895904, gen: 7190, dow: 'Tue' },
    { yr: 2023, mon: 1, dom: 31, hr: 1, imp: 2525323, gen: 5515, dow: 'Tue' },
    { yr: 2023, mon: 1, dom: 31, hr: 2, imp: 2374196, gen: 5450, dow: 'Tue' },
    { yr: 2023, mon: 1, dom: 31, hr: 3, imp: 713464, gen: 7194, dow: 'Tue' },
    { yr: 2023, mon: 1, dom: 31, hr: 4, imp: 624580, gen: 7128, dow: 'Tue' },
    { yr: 2023, mon: 1, dom: 31, hr: 5, imp: 814572, gen: 7132, dow: 'Tue' },
    { yr: 2023, mon: 1, dom: 31, hr: 6, imp: 961806, gen: 7130, dow: 'Tue' },
    { yr: 2023, mon: 1, dom: 31, hr: 7, imp: 1425098, gen: 6830, dow: 'Tue' },
    { yr: 2023, mon: 1, dom: 31, hr: 8, imp: 777800, gep: 163361, gen: 418, dow: 'Tue' },
    { yr: 2023, mon: 1, dom: 31, hr: 9, imp: 352502, exp: 9228, gep: 2975205, h1d: 1369453, dow: 'Tue' },
    { yr: 2023, mon: 1, dom: 31, hr: 10, imp: 2456, exp: 3243905, gep: 5228174, h1d: 755088, dow: 'Tue' },
    { yr: 2023, mon: 1, dom: 31, hr: 11, imp: 240, exp: 2295689, gep: 5539054, dow: 'Tue' },
    { yr: 2023, mon: 1, dom: 31, hr: 12, imp: 118025, exp: 626416, gep: 4886050, h1d: 819404, dow: 'Tue' },
    { yr: 2023, mon: 1, dom: 31, hr: 13, imp: 1586995, exp: 1094158, gep: 3223278, h1d: 561517, dow: 'Tue' },
    { yr: 2023, mon: 1, dom: 31, hr: 14, imp: 323686, exp: 592778, gep: 1879584, dow: 'Tue' },
    { yr: 2023, mon: 1, dom: 31, hr: 15, imp: 669287, gep: 568353, dow: 'Tue' },
    { yr: 2023, mon: 1, dom: 31, hr: 16, imp: 1937473, gep: 149025, dow: 'Tue' },
    { yr: 2023, mon: 1, dom: 31, hr: 17, imp: 2892321, gep: 417, gen: 6352, dow: 'Tue' },
    { yr: 2023, mon: 1, dom: 31, hr: 18, imp: 7396867, gen: 7009, dow: 'Tue' },
    { yr: 2023, mon: 1, dom: 31, hr: 19, imp: 4440413, gen: 7134, dow: 'Tue' },
    { yr: 2023, mon: 1, dom: 31, hr: 20, imp: 2614509, gen: 7127, dow: 'Tue' },
    { yr: 2023, mon: 1, dom: 31, hr: 21, imp: 3651589, gen: 5692, dow: 'Tue' },
    { yr: 2023, mon: 1, dom: 31, hr: 22, imp: 2843383, gen: 6830, dow: 'Tue' },
    { yr: 2023, mon: 1, dom: 31, hr: 23, imp: 2768714, gen: 6950, dow: 'Tue' },
  ]),
}));

const mockMoment = moment('2023-02-13');

jest.mock('../lib/costUtils', () => ({
  initialState: jest.fn().mockReturnValue({
    today: mockMoment,
    selectedDate: mockMoment,
    formattedSelectedDate: '2023-02-13',
    data: [],
    totals: [],
  }),
}));

jest.mock('../lib/dateUtils', () => ({
  ...jest.requireActual('../lib/dateUtils'),
  formatDate: jest.fn().mockReturnValue('2023-02-13'),
}));

jest.mock('../lib/costUtils', () => ({
  recalculateTotals: jest.fn(),
}));

jest.mock('../CustomDatePicker', () => () => <div data-testid="CustomDatePicker-mock" />);
jest.mock('../DailyEnergyUsageLineGraph', () => () => <div data-testid="DailyEnergyUsageLineGraph-mock" />);
jest.mock('../DailyEnergyUsageTable', () => () => <div data-testid="DailyEnergyUsageTable-mock" />);

describe('App', () => {
  it('renders correctly', async () => {
    const selectedDate = moment().subtract(1, 'day');
    const formattedSelectedDate = selectedDate.format('YYYY-MM-DD');

    let renderResult: any;
    await act(async () => {
      renderResult = render(<App />);
    });

    await waitFor(() => {
      expect(pvService.getHourlyUsageDataForDate).toHaveBeenCalledWith(formattedSelectedDate);
    });

    expect(renderResult!.asFragment()).toMatchSnapshot();
  });
});
