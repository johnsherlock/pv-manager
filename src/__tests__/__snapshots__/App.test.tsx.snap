// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders correctly 1`] = `
<DocumentFragment>
  <div
    class="container grid-container"
  >
    <div
      class="row justify-content-center"
    >
      <div
        class="col-12 col-md-9"
      >
        <div
          class="navigation"
        >
          <h1>
            <div
              class="navPrev"
            >
              <div
                class="navigationButton"
              >
                &lt;&lt;
              </div>
            </div>
            <div
              class="date"
            >
              <div
                data-testid="CustomDatePicker-mock"
              />
            </div>
            <div
              class="navNext"
            >
              <div
                class="navigationButton"
              >
                &gt;&gt;
              </div>
            </div>
          </h1>
        </div>
        <div
          data-testid="DailyEnergyUsageTable-mock"
        />
        <div
          class="chart"
        >
          <div
            data-testid="DailyEnergyUsageLineGraph-mock"
          />
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
